pkg/services/accesscontrol/acimpl/service.go

```diff
@@ -335,3 +335,83 @@ func (s *Service) SearchUsersPermissions(ctx context.Context, user *user.SignedI
 
 	return res, nil
 }
+
+func (s *Service) SearchUserPermissions(ctx context.Context, userID, orgID int64, searchOptions accesscontrol.SearchOptions) ([]accesscontrol.Permission, error) {
+	timer := prometheus.NewTimer(metrics.MAccessPermissionsSummary)
+	defer timer.ObserveDuration()
+
+	if permissions, success := s.searchUserPermissionsFromCache(userID, orgID, searchOptions); success {
+		return permissions, nil
+	}
+	return s.searchUserPermissions(ctx, userID, orgID, searchOptions)
+}
+
+func (s *Service) searchUserPermissions(ctx context.Context, userID, orgID int64, searchOptions accesscontrol.SearchOptions) ([]accesscontrol.Permission, error) {
+	// Get permissions for user's basic roles from RAM
+	roleList, err := s.store.GetUsersBasicRoles(ctx, []int64{userID}, orgID)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch basic roles for the user: %w", err)
+	}
+	var roles []string
+	var ok bool
+	if roles, ok = roleList[userID]; !ok {
+		return nil, fmt.Errorf("found no basic roles for user %d in organisation %d", userID, orgID)
+	}
+	permissions := make([]accesscontrol.Permission, 0)
+	for _, builtin := range roles {
+		if basicRole, ok := s.roles[builtin]; ok {
+			for _, permission := range basicRole.Permissions {
+				if PermissionMatchesSearchOptions(permission, searchOptions) {
+					permissions = append(permissions, permission)
+				}
+			}
+		}
+	}
+
+	// Get permissions from the DB
+	dbPermissions, err := s.store.SearchUsersPermissions(ctx, orgID, []int64{userID}, searchOptions)
+	if err != nil {
+		return nil, err
+	}
+	permissions = append(permissions, dbPermissions[userID]...)
+
+	return permissions, nil
+}
+
+func (s *Service) searchUserPermissionsFromCache(userID, orgID int64, searchOptions accesscontrol.SearchOptions) ([]accesscontrol.Permission, bool) {
```

Comment:
can you maybe add a test to check that the cache is correctly populated and used?